reco_analysis/reco_analysis/chatbot/chatbot.py

```diff
@@ -203,11 +203,38 @@ def __init__(
         # Prepare the AI instruction (this acts as guidance for the agent after each run of the chat)
         self.ai_instruct = ai_guidance_doctor if self.role == "Doctor" else ai_guidance_patient
 
+        # Initialize end of conversation flag
+        self.end_conversation = False
+
     def reset(self) -> None:
         """
         Resets the conversation history in memory
         """
         self.memory.clear()
+        self.end_conversation = False
+
+    def get_last_doctor_patient_messages(self) -> typing.Tuple[str, str]:
+        """
+        Retrieves the last doctor and patient messages from the conversation history.
+        
+        Returns:
+            Tuple[str, str]: A tuple containing the last doctor and patient messages.
+        """
+        last_doctor_message = None
+        last_patient_message = None
+
+        if len(self.memory.messages) >= 2 and self.memory.messages[-1].name == 'Patient' and self.memory.messages[-2].name == 'Doctor':
+            last_doctor_message = self.memory.messages[-2].content
+            last_patient_message = self.memory.messages[-1].content
+
+        return last_doctor_message, last_patient_message
+    
+    def detect_and_handle_end(self) -> None:
+        """
+        Detects the end of the conversation based on the last doctor and patient messages.
+        """
+        last_doctor_message, last_patient_message = self.get_last_doctor_patient_messages()
+        self.end_conversation = detect_end(doctor_input=last_doctor_message, patient_input=last_patient_message)
```

Comment:
```suggestion
        last_doctor_message, last_patient_message = self.get_last_doctor_patient_messages()
        if last_doctor_message and last_patient_message:
            self.end_conversation = detect_end(doctor_input=last_doctor_message, patient_input=last_patient_message)
```

because what's returned could be None's.